### PR TITLE
Ignore cancellation error when search is cancelled

### LINKED ISSUE #63702: AsyncSearchActionIT#testSearchPhaseFailureNoCause fails
This test failure can be caused by the changes in https://github.com/elastic/elasticsearch/pull/63520.

Some build scans: 
  - https://gradle-enterprise.elastic.co/s/cme2mjkkijb3u

I will mute the test and take a look at the failure.

### LINKED ISSUE #64012: [CI] Failures in update_by_query/35_search_failure/Response format for search failures
**Build scan**:
https://gradle-enterprise.elastic.co/s/6oq2ek5easfsm

**Repro line**:

./gradlew ':modules:reindex:yamlRestTest' --tests "org.elasticsearch.index.reindex.ReindexClientYamlTestSuiteIT.test {yaml=update_by_query/35_search_failure/Response format for search failures}" \
  -Dtests.seed=5ADD9729CFD6E795 \
  -Dtests.security.manager=true \
  -Dtests.locale=fi-FI \
  -Dtests.timezone=Europe/Helsinki \
  -Druntime.java=11

**Reproduces locally?**:
No

**Applicable branches**:
master

**Failure history**:
Looks like this only fails on master so far (aprox. six times since Oct 14th so far)

**Failure excerpt**:
```
org.elasticsearch.index.reindex.ReindexClientYamlTestSuiteIT > test {yaml=update_by_query/35_search_failure/Response format for search failures} FAILED
    java.lang.AssertionError: Failure at [update_by_query/35_search_failure:18]: expected [400] status code but api [update_by_query] returned [500 Internal Server Error] [---
    error:
      root_cause:
      - type: "script_exception"
        reason: "runtime error"
        script_stack:
        - "throw new IllegalArgumentException(\"Cats!\")"
        - "      ^---- HERE"
        script: "throw new IllegalArgumentException(\"Cats!\")"
        lang: "painless"
        position:
          offset: 6
          start: 0
          end: 43
        stack_trace: "org.elasticsearch.script.ScriptException: runtime error\n\tat org.elasticsearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:96)\n\
          \tat org.elasticsearch.painless.PainlessScript$Script.execute(throw new IllegalArgumentException(\"\
          Cats!\"):1)\n\tat org.elasticsearch.index.query.ScriptQueryBuilder$ScriptQuery$1$1.matches(ScriptQueryBuilder.java:192)\n\
          \tat org.apache.lucene.search.Weight$DefaultBulkScorer.scoreRange(Weight.java:259)\n\
          \tat org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:239)\n\
          \tat org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:56)\n\
          \tat org.apache.lucene.search.BulkScorer.score(BulkScorer.java:39)\n\tat org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:227)\n\
          \tat org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:200)\n\
          \tat org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:445)\n\
          \tat org.elasticsearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:345)\n\
          \tat org.elasticsearch.search.query.QueryPhase.executeInternal(QueryPhase.java:296)\n\
          \tat org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:148)\n\
          \tat org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:374)\n\
          \tat org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:433)\n\
          \tat org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:403)\n\
          \tat org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58)\n\
          \tat org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73)\n\
          \tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)\n\
          \tat org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44)\n\
          \tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:733)\n\
          \tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)\n\
          \tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\
          \tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\
          \tat java.base/java.lang.Thread.run(Thread.java:834)\nCaused by: java.lang.IllegalArgumentException:\
          \ Cats!\n\tat org.elasticsearch.painless.PainlessScript$Script.execute(throw\
          \ new IllegalArgumentException(\"Cats!\"):7)\n\t... 23 more\n"
      - type: "task_cancelled_exception"
        reason: "The parent task was cancelled, shouldn't start any child tasks"
        stack_trace: "org.elasticsearch.tasks.TaskCancelledException: The parent task\
          \ was cancelled, shouldn't start any child tasks\n\tat org.elasticsearch.tasks.TaskManager$CancellableTaskHolder.registerChildNode(TaskManager.java:560)\n\
          \tat org.elasticsearch.tasks.TaskManager.registerChildNode(TaskManager.java:255)\n\
          \tat org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:570)\n\
          \tat org.elasticsearch.transport.TransportService.sendChildRequest(TransportService.java:653)\n\
          \tat org.elasticsearch.transport.TransportService.sendChildRequest(TransportService.java:645)\n\
          \tat org.elasticsearch.action.search.SearchTransportService.sendExecuteQuery(SearchTransportService.java:146)\n\
          \tat org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.executePhaseOnShard(SearchQueryThenFetchAsyncAction.java:83)\n\
          \tat org.elasticsearch.action.search.AbstractSearchAsyncAction.lambda$performPhaseOnShard$3(AbstractSearchAsyncAction.java:245)\n\
          \tat org.elasticsearch.action.search.AbstractSearchAsyncAction.performPhaseOnShard(AbstractSearchAsyncAction.java:280)\n\
          \tat org.elasticsearch.action.search.AbstractSearchAsyncAction.run(AbstractSearchAsyncAction.java:216)\n\
          \tat org.elasticsearch.action.search.AbstractSearchAsyncAction.executePhase(AbstractSearchAsyncAction.java:366)\n\
          \tat org.elasticsearch.action.search.AbstractSearchAsyncAction.start(AbstractSearchAsyncAction.java:183)\n\
          \tat org.elasticsearch.action.search.TransportSearchAction$5.run(TransportSearchAction.java:790)\n\
[...]
```

The tests expects an exception but it doesn't seem to be the right one in this case.

### PR DESCRIPTION
**Pull request by dnhatn**
Since #63520, we will cancel a search that hits shard failures and does not accept partial results. However, that change can return the wrong HTTP code for bad requests (from 4xx to 5xx) due to the cancellation.

Relates #63520
Closes #64012 
Closes #63702

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java [hunk 1]
```diff
@@ -438,7 +439,7 @@ protected void onShardGroupFailure(int shardIndex, SearchShardTarget shardTarget
     @Override
     public final void onShardFailure(final int shardIndex, @Nullable SearchShardTarget shardTarget, Exception e) {
         // we don't aggregate shard failures on non active shards (but do keep the header counts right)
-        if (TransportActions.isShardNotAvailableException(e) == false) {
+        if (TransportActions.isShardNotAvailableException(e) == false && (requestCancelled.get() && isTaskCancelledException(e)) == false) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:442]: Can you update the comment above to say that we don't aggregate shard failures on search cancelled internally ?

**Review by jimczi — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:443]: ++, added in https://github.com/elastic/elasticsearch/pull/64240/commits/cd53c74a057a1bcddfe7d7429643607f993312cf

**Comment by dnhatn** [comment]
Thanks Jim!

**Pull request merged**